### PR TITLE
Update `Cppyy::GetScope` to handle nested namespaces and templates

### PR DESCRIPTION
**Pull request by Vipul-Cariappa**
11 new tests pass.
1 xfail, segfaults with this change.

### CODE REVIEW

File: clingwrapper/src/clingwrapper.cxx [hunk 1]
```diff
@@ -571,7 +571,32 @@ Cppyy::TCppScope_t Cppyy::GetScope(const std::string& name,
                                  + name + "'\n");
 #endif // NDEBUG
 
-    return Cpp::GetScope(name, parent_scope);
+    Cppyy::TCppScope_t scope = Cpp::GetScope(name, parent_scope);
+    if (!scope && (parent_scope == nullptr || parent_scope == Cpp::GetGlobalScope()))
+        scope = Cpp::GetScopeFromCompleteName(name);
```
Comment (Vipul-Cariappa) [inline clingwrapper/src/clingwrapper.cxx:576]: I was thinking about updating
```c++
TCppScope_t Cpp::GetScopeFromCompleteName(const std::string &name)
```
to 
```c++
TCppScope_t Cpp::GetScopeFromCompleteName(const std::string &name, TCppScope_t parent)
```
at CppInterOp.

Any suggestions?

Comment (aaronj0) [reply clingwrapper/src/clingwrapper.cxx:576]: Yes, but I think the reason for this interface was to provide the fully scoped name. Do we benefit from providing the parent scope too?

**Comment by aaronj0** [comment]
Looks good! We should probably prevent the seg fault on the xfailing test though.. does it come from the scope lookup or the template instantiation?

**Comment by Vipul-Cariappa** [comment]
> Looks good! We should probably prevent the seg fault on the xfailing test though.. does it come from the scope lookup or the template instantiation? Maybe a case we can address in clingwrapper/ or whatever calls `Cppyy::GetScope` in CPyCppyy

I suspect it might be after the call to `Cppyy::GetScope`.

File: clingwrapper/src/clingwrapper.cxx (hunk 1 above)
Comment (Vipul-Cariappa) [reply clingwrapper/src/clingwrapper.cxx:576]: `Cpp::GetScope` does not handle namespaces, like `my_namespace::my_class`.
And `Cpp::GetScopeFromCompleteName` does not have the parent scope argument.

The code here will be simplified. Thats a benefit.

**Comment by aaronj0** [comment]
> > Looks good! We should probably prevent the seg fault on the xfailing test though.. does it come from the scope lookup or the template instantiation? Maybe a case we can address in clingwrapper/ or whatever calls `Cppyy::GetScope` in CPyCppyy
> 
> I suspect it might be after the call to `Cppyy::GetScope`.

Yes, I meant to ask within clingwrapper does it happen at `Cpp::GetScope` or `Cpp::InstantiateTemplateFunction`. We are either doing a faulty scope lookup or template instantiation and we should catch what is causing this

**Comment by Vipul-Cariappa** [comment]
Hi @aaronj0. I investigated the failing test.
Test `test03_stllike_preinc` fails while compiling the following code snippet
```c++
namespace __cppyy_internal {
void init_PreIncrement__Iterator(PreIncrement::Iterator*& self, const Token& current) {
  self = new PreIncrement::Iterator{current};
} }
```
The stack trace goes through Pythonize.cxx:1779.

If I run that single test standalone, it passes.

Performing `dump()` on the `__cppyy_internal` namespace I get the following:

With pytest:
Screenshot From 2024-12-03 11-49-01

Running standalone:
Screenshot From 2024-12-03 11-48-42

You can see that the second `ParamVarDecl` is different.

ping @vgvassilev

**Comment by vgvassilev** [comment]
I think it is fine to make it from xfail a crash and investigate later.

File: clingwrapper/src/clingwrapper.cxx [hunk 2]
```diff
@@ -571,7 +571,32 @@ Cppyy::TCppScope_t Cppyy::GetScope(const std::string& name,
...
+    if (scope)
+        return scope;
```
Comment (vgvassilev) [inline clingwrapper/src/clingwrapper.cxx:578]:
```suggestion
    if (Cppyy::TCppScope_t scope = Cpp::GetScope(name, parent_scope))
      return scope;

    if (!parent_scope || parent_scope == Cpp::GetGlobalScope())
      if (Cppyy::TCppScope_t scope = Cpp::GetScopeFromCompleteName(name))
         return scope;
```

You can simplify the code using that pattern...

File: clingwrapper/src/clingwrapper.cxx [hunk 3]
```diff
@@ -565,13 +565,31 @@ Cppyy::TCppScope_t Cppyy::GetUnderlyingScope(TCppScope_t scope)
...
+    if (name.find('<') != std::string::npos) {
+      // Templated Type; May need instantiation
+      size_t start = name.find('<');
+      size_t end = name.rfind('>');
+      std::string params = name.substr(start + 1, end - start - 1);
+
+      std::string pure_name = name.substr(0, start);
+      Cppyy::TCppScope_t scope = Cpp::GetScope(pure_name, parent_scope);
+      if (!scope && (!parent_scope || parent_scope == Cpp::GetGlobalScope()))
+        scope = Cpp::GetScopeFromCompleteName(pure_name);
```
Comment (vgvassilev) [inline clingwrapper/src/clingwrapper.cxx:583]:
```suggestion
      Cppyy::TCppScope_t scope = Cpp::GetScope(pure_name, parent_scope)
      if (scope)
        return scope;
      
      if (!parent_scope || parent_scope == Cpp::GetGlobalScope())
        scope = Cpp::GetScopeFromCompleteName(pure_name);
```

Comment (Vipul-Cariappa) [reply clingwrapper/src/clingwrapper.cxx:583]: We are not returning the `scope` here. We are using `scope` to `InstantiateTemplate` and returning the template.

Comment (vgvassilev) [reply clingwrapper/src/clingwrapper.cxx:583]: ah, ok, good point..

**Pull request merged**